circleguard/investigator.py

```diff
@@ -226,41 +242,146 @@ def median_frametime(frametimes):
+                # exception for sliders/spinners, where notelock ends after hitobject end time if it's earlier
+                if hitobj_type != 0:
+                    notelock_end_time = min(notelock_end_time, hitobj_end_time)
+            # after sliderbug fix, notelock ends after hitobject end time
+            else:
+                notelock_end_time = hitobj_end_time
+                # apparently notelock was increased by 2ms for circles (from testing)
+                if hitobj_type == 0:
+                    notelock_end_time += hitwindow + 2
+                # account for 0 frames that cause notelock to be 1ms shorter
+                if replay_data[press_i][2]:
+                    notelock_end_time -= 1
+
+
+            # can't press on hitobjects before hitwindowmiss
+            if press_time < hitobj_time - 399.5:
                 press_i += 1
+                continue
+
+            if press_time < hitobj_time - hitwindow:
+                # pressing on a circle or slider during hitwindowmiss will cause a miss
+                if np.linalg.norm(press_xy - hitobj_xy) <= hitradius and hitobj_type != 2:
+
+                    # sliders don't disappear after missing
+                    # so we skip to the press_i that is after notelock_end_time
+                    press_i += 1
+                    if press_i < len(replay_data) and hitobj_type == 1 and version >= Investigator.VERSION_SLIDERBUG_FIXED:
+                        while replay_data[press_i][0] <= notelock_end_time:
+                            press_i += 1
+                            if press_i >= len(replay_data):
+                                break
+                    object_i += 1
+                # keypress not on object, so we move to the next keypress
+                else:
+                    press_i += 1
+            elif press_time >= notelock_end_time:
+                # can no longer interact with hitobject after notelock_end_time
+                # so we move to the next object
                 object_i += 1
+            else:
+                if press_time < hitobj_time + hitwindow and np.linalg.norm(press_xy - hitobj_xy) <= hitradius and hitobj_type != 2:
+                    array.append([hitobj_time, press_time])
+
+                    # sliders don't disappear after clicking
+                    # so we skip to the press_i that is after notelock_end_time
+                    press_i += 1
+                    if press_i < len(replay_data) and hitobj_type == 1 and version >= Investigator.VERSION_SLIDERBUG_FIXED:
+                        while replay_data[press_i][0] <= notelock_end_time:
+                            press_i += 1
+                            if press_i >= len(replay_data):
+                                break
+                    object_i += 1
+                # keypress not on object, so we move to the next keypress
+                else:
+                    press_i += 1
 
         return array
```

Comment:
I expect hits to be useful for things other than UR, so we should make a new class `Hit` for it which contents the replay data at that frame (or equivalent information in the form of t, xy, and k), the hitobject that got hit, and maybe anything else that's relevant

This is more of a note for myself but if you get to it first feel free